File: dotnet/cafdemalihapa/Applications/Something.cs
using System.Text;
using System.Text.Json;
using cafdemalihapa.Directories.Repository.Dotnet.Cafdemalihapa.Scripts.EnvironmentVariablesSource.SeparationFilement;
using cafdemalihapa.Directories.Repository.Dotnet.Cafdemalihapa.Scripts.EnvironmentVariablesSource.Files.Jsons;

namespace cafdemalihapa.Applications
{
    public class Something(
            PersistentVariablesFile persistentVariablesFile,
            MutantVariablesFile mutantVariablesFile
        )
    {
        public static Dictionary<string, string> PairUpVariablesWithTheirValue(
            string fileNamePath,
            Dictionary<string, string> environmentVariablesSourceDictionary
        )
        {
            Dictionary<string, string> fileContentDictionaryToWriteToFile = [];

            const int bufferSize = 128;
            using var fileStream = File.OpenRead(fileNamePath);
            using var streamReader = new StreamReader(fileStream, Encoding.UTF8, true, bufferSize);

            while (streamReader.ReadLine() is { } line)
            {
                var brokenLine = line.Split("=");
                var key = brokenLine[0];
                var value = brokenLine[1];
                _ = environmentVariablesSourceDictionary.TryGetValue(key, out var val);
                fileContentDictionaryToWriteToFile.Add(key, val ?? "");
            }

            return fileContentDictionaryToWriteToFile;
        }

        public static string ToUnderscoreCase(string str)
        {
            return string.Concat(str.Select((x, i) => i > 0 && char.IsUpper(x) ? "_" + x.ToString() : x.ToString())).ToLower();
        }

        private static Dictionary<string, string> ReadKeyValueFromJsonFile<T>(string path)
        {
            Dictionary<string, string> fileContentDictionary = [];

            using StreamReader r = new(path);
            var rawJsonData = r.ReadToEnd();

            var variables = Newtonsoft.Json.JsonConvert
                .DeserializeObject<T>(rawJsonData);

            var variablesSerialized = JsonSerializer.Serialize(variables);
            var dict = Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, string>>(variablesSerialized);
            foreach (var (key, value) in from kv in dict
                                         let key = ToUnderscoreCase(kv.Key).ToUpper()
                                         let value = kv.Value
                                         select (key, value))
            {
                fileContentDictionary.Add(key, value);
            }

            return fileContentDictionary;
        }

        public Dictionary<string, string> PairUpEnvironmentVariablesSeparationFilement()
        {
            var mutantVariablesFilePath = mutantVariablesFile.GetPath();
            var mutantVariablesDictionary =
                ReadKeyValueFromJsonFile<MutantVariables>(mutantVariablesFilePath);

            var persistentVariablesFilePath = persistentVariablesFile.GetPath();
            var persistentVariablesDictionary =
                ReadKeyValueFromJsonFile<PersistentVariables>(persistentVariablesFilePath);

            var environmentVariablesConcatenated = mutantVariablesDictionary
                .Concat(persistentVariablesDictionary)
                .ToDictionary(entry => entry.Key, entry => entry.Value);

            return environmentVariablesConcatenated;
        }
    }
}

File: dotnet/cafdemvers/zzq-cafdemwimu/cafdemwimu.console/Applications/Something.fs
namespace cafdemwimu.console.Applications

open System
open System.IO
open System.Text
open System.Text.Json
open System.Linq
open System.Collections.Generic
open Microsoft.Extensions.Logging
open cafdemwimu.console.Directories.Repository.Dotnet.Scripts.EnvironmentVariablesSource
open cafdemwimu.console.Directories.Repository.Dotnet.Scripts.EnvironmentVariablesSource.Files.Jsons

type Something
    (
        persistentVariablesFile: PersistentVariablesFile,
        mutantVariablesFile: MutantVariablesFile,
        logger: ILogger<Something>
    ) =

    static member PairUpVariablesWithTheirValue
        (fileNamePath: string, environmentVariablesSourceDictionary: Dictionary<string, string>)
        =
        let fileContentDictionaryToWriteToFile = Dictionary<string, string>()

        let bufferSize = 128
        use fileStream = File.OpenRead(fileNamePath)
        use streamReader = new StreamReader(fileStream, Encoding.UTF8, true, bufferSize)

        let mutable line = streamReader.ReadLine()

        while not (isNull line) do
            let brokenLine = line.Split('=')
            let key = brokenLine.[0]
            let _value = brokenLine.[1]
            let _, value = environmentVariablesSourceDictionary.TryGetValue key
            fileContentDictionaryToWriteToFile.Add(key, (if isNull value then "" else value))
            line <- streamReader.ReadLine()

        fileContentDictionaryToWriteToFile

    static member ToUnderscoreCase(str: string) =
        str
        |> Seq.mapi (fun i x -> if i > 0 && Char.IsUpper x then "_" + string x else string x)
        |> String.concat ""

        |> fun s -> s.ToLower()

    static member private ReadKeyValueFromJsonFile<'T>(path: string) =
        let fileContentDictionary = Dictionary<string, string>()

        use r = new StreamReader(path)
        let rawJsonData = r.ReadToEnd()

        let options = JsonSerializerOptions(PropertyNameCaseInsensitive = true)
        let variables = JsonSerializer.Deserialize<'T>(rawJsonData, options)

        let variablesSerialized = JsonSerializer.Serialize(variables)
        use document = JsonDocument.Parse(variablesSerialized)

        // Flatten the (possibly nested) object into UPPER_SNAKE keys, using leaf
        // property names so nested groups (e.g. "PrimaryApplication") still produce
        // the flat env keys (PRIMARY_APPLICATION_LOCATION, ...) that the templates expect.
        let rec flatten (element: JsonElement) =
            for property in element.EnumerateObject() do
                match property.Value.ValueKind with
                | JsonValueKind.Object -> flatten property.Value
                | _ ->
                    let key = (Something.ToUnderscoreCase property.Name).ToUpper()

                    let value =
                        match property.Value.ValueKind with
                        | JsonValueKind.Null -> ""
                        | JsonValueKind.True -> "true"
                        | JsonValueKind.False -> "false"
                        | JsonValueKind.String -> property.Value.GetString()
                        | _ -> property.Value.ToString()

                    fileContentDictionary.[key] <- value

        flatten document.RootElement

        fileContentDictionary

    member _.PairUpEnvironmentVariablesSeparationFilement() =
        let mutantVariablesFilePath: string = mutantVariablesFile.GetPath()
        let persistentVariablesFilePath: string = persistentVariablesFile.GetPath()

        let mutantVariablesDictionary: Dictionary<string, string> =
            Something.ReadKeyValueFromJsonFile<MutantVariables>(mutantVariablesFilePath)


        let persistentVariablesDictionary: Dictionary<string, string> =
            Something.ReadKeyValueFromJsonFile<PersistentVariables>(persistentVariablesFilePath)

        let environmentVariablesConcatenated: Dictionary<string, string> =
            mutantVariablesDictionary
                .Concat(persistentVariablesDictionary)
                .ToDictionary((fun (entry: KeyValuePair<string, string>) -> entry.Key), (fun entry -> entry.Value))

        environmentVariablesConcatenated
